src/Compilers/Core/Portable/Operations/CustomDataFlowAnalysis.cs

```diff
@@ -0,0 +1,402 @@
+            ImmutableArray<TBasicBlock> blocks,
+            TDataFlowAnalyzer analyzer,
+            int firstBlockOrdinal,
+            int lastBlockOrdinal,
+            TBlockAnalysisData initialAnalysisData,
+            ArrayBuilder<TBasicBlock> unreachableBlocksToVisit,
+            SortedSet<int> outOfRangeBlocksToVisit,
+            PooledDictionary<ControlFlowRegion, bool> continueDispatchAfterFinally,
+            PooledHashSet<ControlFlowRegion> dispatchedExceptionsFromRegions,
+            CancellationToken cancellationToken)
+        {
+            var toVisit = new SortedSet<int>();
+
+            var firstBlock = blocks[firstBlockOrdinal];
+            analyzer.SetCurrentAnalysisData(firstBlock, initialAnalysisData);
+            toVisit.Add(firstBlock.Ordinal);
+
+            var processedBlocks = PooledHashSet<TBasicBlock>.GetInstance();
+            TBlockAnalysisData resultAnalysisData = default;
+
+            do
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                TBasicBlock current;
+                if (toVisit.Count > 0)
+                {
+                    var min = toVisit.Min;
+                    toVisit.Remove(min);
+                    current = blocks[min];
+                }
+                else
+                {
+                    int index;
+                    current = null;
+                    for (index = 0; index < unreachableBlocksToVisit.Count; index++)
+                    {
+                        var unreachableBlock = unreachableBlocksToVisit[index];
+                        if (unreachableBlock.Ordinal >= firstBlockOrdinal && unreachableBlock.Ordinal <= lastBlockOrdinal)
+                        {
+                            current = unreachableBlock;
+                            break;
+                        }
+                    }
+
+                    if (current == null)
+                    {
+                        continue;
+                    }
+
+                    unreachableBlocksToVisit.RemoveAt(index);
+                    if (processedBlocks.Contains(current))
+                    {
+                        // Already processed from a branch from another unreachable block.
+                        continue;
+                    }
+
+                    analyzer.SetCurrentAnalysisData(current, analyzer.GetEmptyAnalysisData());
+                }
+
+                if (current.Ordinal < firstBlockOrdinal || current.Ordinal > lastBlockOrdinal)
+                {
+                    outOfRangeBlocksToVisit.Add(current.Ordinal);
+                    continue;
+                }
+
+                if (current.Ordinal == current.EnclosingRegion.FirstBlockOrdinal)
+                {
+                    // We are revisiting first block of a region, so we need to again dispatch exceptions from region.
+                    dispatchedExceptionsFromRegions.Remove(current.EnclosingRegion);
```

Comment:
This was added as we might visit same block/region multiple times during fixed point iteration, and we should do the dispatch every time we start analyzing a region from its first block.